arangod/Replication2/ReplicatedState/LeaderStateManager.tpp

```diff
@@ -130,31 +135,60 @@ LeaderStateManager<S>::LeaderStateManager(
     std::shared_ptr<ReplicatedState<S>> const& parent,
     std::shared_ptr<replicated_log::ILogLeader> leader,
     std::unique_ptr<ReplicatedStateCore> core,
+    std::unique_ptr<ReplicatedStateToken> token,
     std::shared_ptr<Factory> factory) noexcept
     : parent(parent),
       logLeader(std::move(leader)),
       internalState(LeaderInternalState::kWaitingForLeadershipEstablished),
       core(std::move(core)),
-      factory(std::move(factory)) {}
+      token(std::move(token)),
+      factory(std::move(factory)) {
+  TRI_ASSERT(this->core != nullptr);
+  TRI_ASSERT(this->token != nullptr);
+}
 
 template<typename S>
 auto LeaderStateManager<S>::getStatus() const -> StateStatus {
+  if (token == nullptr || core == nullptr) {
+    TRI_ASSERT(core == nullptr && token == nullptr);
+    THROW_ARANGO_EXCEPTION(
+        TRI_ERROR_REPLICATION_REPLICATED_LOG_FOLLOWER_RESIGNED);
+  }
   LeaderStatus status;
-  status.log = std::get<replicated_log::LeaderStatus>(
-      logLeader->getStatus().getVariant());
-  status.state.state = internalState;
-  status.state.lastChange = lastInternalStateChange;
+  status.managerState.state = internalState;
+  status.managerState.lastChange = lastInternalStateChange;
   if (internalState == LeaderInternalState::kRecoveryInProgress &&
       recoveryRange) {
-    status.state.detail = "recovery range is " + to_string(*recoveryRange);
+    status.managerState.detail =
+        "recovery range is " + to_string(*recoveryRange);
   } else {
-    status.state.detail = std::nullopt;
+    status.managerState.detail = std::nullopt;
   }
+  status.snapshot = token->snapshot;
+  status.generation = token->generation;
   return StateStatus{.variant = std::move(status)};
 }
 
 template<typename S>
-auto LeaderStateManager<S>::getSnapshotStatus() const -> SnapshotStatus {
-  return core->snapshot;
+auto LeaderStateManager<S>::resign() && noexcept
+    -> std::pair<std::unique_ptr<ReplicatedStateCore>,
+                 std::unique_ptr<ReplicatedStateToken>> {
+  LOG_TOPIC("edcf3", TRACE, Logger::REPLICATED_STATE)
+      << "Leader manager resign";
+  TRI_ASSERT(core != nullptr);
+  TRI_ASSERT(token != nullptr);
+  return {std::move(core), std::move(token)};
+}
+
+template<typename S>
+void LeaderStateManager<S>::waitForParticipantResigned() {
+  logLeader->waitFor(LogIndex{std::numeric_limits<std::uint64_t>::max()})
+      .thenFinal([weak = this->weak_from_this()](auto&&) {
+        if (auto self = weak.lock(); self != nullptr) {
+          if (auto parentPtr = self->parent.lock(); parentPtr != nullptr) {
+            parentPtr->forceRebuild();
+          }
+        }
+      });
```

Comment:
`waitForParticipantResigned` doesn't actually wait, but it also doesn't return the future. Maybe we should rename it to avoid confusion?